### PR TITLE
FTP.ls - Adding a method to prune the tree at "source."

### PR DESCRIPTION
**Pull request by mpricope**
Only needed for folders. Individual files can be filtered after emiting
without much overhead.
Folders are expensive to be filtered afterwards.

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @mpricope,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by mpricope** [comment]
Issue #452

File: ftp/src/main/scala/akka/stream/alpakka/ftp/javadsl/FtpApi.scala [hunk 1]
```diff
@@ -91,7 +91,25 @@ sealed trait FtpApi[FtpClient] { _: FtpSourceFactory[FtpClient] =>
...
+   *                       if you want to enter that remote folder.
+   *                       Default behaviour is full recursiv which is equivalent with calling this function
+   *                       with [[ls(basePath,connectionSettings,f->true)]].
+   *
+   *                       Calling [[ls(basePath,connectionSettings,f->false)]] will emit only the files and folder in
+   *                       non-recursive fashion
+   *
+   * @return A [[Source]] of [[FtpFile]]s
+   */
+  def ls(basePath: String, connectionSettings: S, branchSelector: FtpFile => Boolean): Source[FtpFile, NotUsed] =
```
Comment (johanandren) [inline ftp/src/main/scala/akka/stream/alpakka/ftp/javadsl/FtpApi.scala]: Scala function for Java is only possible for Scala 2.12 but we support both 2.11 and 2.12 so this has to accept a `java.util.function.Function`

Comment (mpricope) [reply ftp/src/main/scala/akka/stream/alpakka/ftp/javadsl/FtpApi.scala]: thanks for the feedback .... I will update this and resubmit

**Pull request merged**